Documentation/design-docs/host-startup-hook.md

```diff
@@ -0,0 +1,96 @@
+# Host startup hook
+
+For .NET Core 3+, we want to provide a low-level hook that allows
+injecting managed code to run before the main application's entry
+point. This hook will make it possible for the host to customize the
+behavior of managed applications during process launch, after they
+have been deployed.
+
+## Motivation
+
+This would allow hosting providers to define custom configuration and
+policy in managed code, including settings that potentially influence
+load behavior of the main entry point such as the
+`AssemblyLoadContext` behavior. The hook could be used to set up
+tracing or telemetry injection, to set up callbacks for handling
+Debug.Assert (if we make such an API available), or other
+environment-dependent behavior. The hook is separate from the entry
+point, so that user code doesn't need to be modified.
+
+## Proposed behavior
+
+The `DOTNET_STARTUP_HOOKS` environment variable can be used to specify
+a list of managed assemblies and type names that contain `public
+static void Initialize()` methods, which will be called in the order
+specified, before the `Main` entry point:
+
+```
+DOTNET_STARTUP_HOOKS=/path/to/StartupHook1.dll!StartupHookNamespace.StartupHookType1;/path/to/StartupHook2.dll!StartupHookNamespace.StartupHookType2
+```
+
+This variable is a list of absolute assembly paths, each with a type
+name following an exclamation mark. The list is delimited by the
+platform-specific path separator (`;` on Windows and `:` on Unix). It
```

Comment:
I don't think we are well served by making things platform specific.   Basically it takes something that SHOULD have been platform neutral (this startup variable) and requires users of this feature to make platform specific adaptations.   We should not introduce gratuitous platform differences.  It just leads to unnecessary  complexity for your users. 

The simplest resolution to this is to simply pick one (it becomes the 'dotnet' convention for separating things).  I would recommend ';' because ':' will be needed in most windows paths, but the reverse is not true.    If you REALLY feel strongly you can make the UNIX version support both ';' and ':' but would not bother here.  This is especially true because we expect in the common case it will have only one entry.

Reply:
> The simplest resolution to this is to simply pick on

I agree with you if this is in the config file.

For environment variables, it is an interesting question whether to be platform native or unified. My experience is that the separators are platform native for environment variables. For example, here is what Java `CLASSPATH` does on Windows vs. Unix:

https://docs.oracle.com/javase/6/docs/technotes/tools/windows/classpath.html
https://docs.oracle.com/javase/8/docs/technotes/tools/unix/classpath.html

Are there examples where folks choose to be platform unified for path separators in environment variables?

Reply:
Hmmm ... maybe we should not be looking at these as paths (they are not actually proper paths because they have type appended), and so it is not worth worrying about being platform native?

Reply:
I think there's a reasonable argument in either direction. Consistency with existing PATH-like variables is nice. OTOH, they're not really paths, so consistency with PATH-like variables may not be essential. Another consideration is that platform-specific separators make it easy to set PATHs on in bash without quotes:
```
PATH=a:b
```
works on unix, whereas
```
PATH=a;b
```
gets split into two commands unless the `"a;b"` is quoted. That's an argument for using platform-specific separators in PATH-like variables... *but* in our case `!` causes problems in bash anyway, even if it's quoted, because bash interprets the `!` as a history expansion prefix:
```
$ DOTNET_STARTUP_HOOKS="path!type"
-bash: !type": event not found
```
Setting this requires single quotes:
```
DOTNET_STARTUP_HOOKS='path!type'
```
I think that pretty much defeats the purpose (from a usability perspective) of having platform-specific separators. So at this point I'd suggest either sticking with `;` or using something else (`|`?) instead of `!` to prevent the need for single quotes on unix. Note that even `|` as a path/type separator would require double quotes.

Reply:
I was against the idea of having a hard-coded default type name just because it seemed like too much magic. However, if it wasn't a default, but rather the **required** name of the type, I think that is entirely reasonable.

Reply:
> Also I think you will find that we use ';' as a separator in other places already

Do you have a example? I believe that the .NET APIs actually lead you to using platform native path separator in environment variables because of we have `Path.PathSeparator` since forever. In other words, the platform specific shim has been created for you already.

> hard-coded default type name 

+1 The configurability of type name seems to have net negative value. It just requires users to do more typing in vast majority of cases. A required hardcoded type name should be just fine.

Reply:
All right, so it sounds like we're leaning towards platform-specific separators, without the type name in the syntax. I'll specify that the type **must** be named "StartupHook" (in the global namespace). This will allow the hook to be set without quotes on unix and windows:
```
DOTNET_STARTUP_HOOKS=/path/to/startupHook.dll:/path/to/startupHook2.dll
```
and on windows:
```
set DOTNET_STARTUP_HOOKS=d:\path\to\startupHook.dll;d:\path\to\startupHook2.dll
```